File: SteganographyProcessor.java
import java.awt.Graphics;
import java.awt.image.BufferedImage;

public class SteganographyProcessor {

    public static BufferedImage encode(BufferedImage img, String message) {
        String fullMessage = message + "#"; // end marker
        byte[] bytes = fullMessage.getBytes();
        int msgIndex = 0, bitIndex = 0;

        BufferedImage result = new BufferedImage(img.getWidth(), img.getHeight(), BufferedImage.TYPE_INT_RGB);
        Graphics g = result.getGraphics();
        g.drawImage(img, 0, 0, null);
        g.dispose();

        outer:
        for (int y = 0; y < result.getHeight(); y++) {
            for (int x = 0; x < result.getWidth(); x++) {
                if (msgIndex >= bytes.length) break outer;

                int rgb = result.getRGB(x, y);
                int r = (rgb >> 16) & 0xFF;
                int gVal = (rgb >> 8) & 0xFF;
                int b = rgb & 0xFF;

                int bit = (bytes[msgIndex] >> (7 - bitIndex)) & 1;
                b = (b & 0xFE) | bit;

                int newRGB = (r << 16) | (gVal << 8) | b;
                result.setRGB(x, y, newRGB);

                bitIndex++;
                if (bitIndex == 8) {
                    bitIndex = 0;
                    msgIndex++;
                }
            }
        }

        return result;
    }

    public static String decode(BufferedImage img) {
        StringBuilder message = new StringBuilder();
        int currentByte = 0;
        int bitCount = 0;

        outer:
        for (int y = 0; y < img.getHeight(); y++) {
            for (int x = 0; x < img.getWidth(); x++) {
                int rgb = img.getRGB(x, y);
                int b = rgb & 0xFF;

                currentByte = (currentByte << 1) | (b & 1);
                bitCount++;

                if (bitCount == 8) {
                    char ch = (char) currentByte;
                    if (ch == '#') break outer;
                    message.append(ch);
                    currentByte = 0;
                    bitCount = 0;
                }
            }
        }

        return message.toString();
    }
}

File: steganography_processor.py
from PIL import Image

def encode(img, message):
    full_message = message + "#"
    bytes_data = full_message.encode()
    msg_index = 0
    bit_index = 0

    img = img.convert("RGB")
    pixels = img.load()

    width, height = img.size

    for y in range(height):
        for x in range(width):
            if msg_index >= len(bytes_data):
                return img

            r, g, b = pixels[x, y]

            bit = (bytes_data[msg_index] >> (7 - bit_index)) & 1
            b = (b & 0xFE) | bit

            pixels[x, y] = (r, g, b)

            bit_index += 1
            if bit_index == 8:
                bit_index = 0
                msg_index += 1

    return img


def decode(img):
    img = img.convert("RGB")
    pixels = img.load()

    width, height = img.size

    message = ""
    current_byte = 0
    bit_count = 0

    for y in range(height):
        for x in range(width):
            r, g, b = pixels[x, y]

            current_byte = (current_byte << 1) | (b & 1)
            bit_count += 1

            if bit_count == 8:
                char = chr(current_byte)
                if char == "#":
                    return message
                message += char
                current_byte = 0
                bit_count = 0

    return message